server/src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -0,0 +1,400 @@
+    }
+
+    @Transactional
+    public boolean isProductMappedToOwner(Product product, Owner owner) {
+        String jpql = "SELECT count(op) FROM OwnerProduct op " +
+            "WHERE op.owner.id = :owner_id AND op.product.uuid = :product_uuid";
+
+        long count = (Long) this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .setParameter("product_uuid", product.getUuid())
+            .getSingleResult();
+
+        return count > 0;
+    }
+
+    @Transactional
+    public boolean mapProductToOwner(Product product, Owner owner) {
+        if (!this.isProductMappedToOwner(product, owner)) {
+            this.create(new OwnerProduct(owner, product));
+
+            return true;
+        }
+
+        return false;
+    }
+
+    @Transactional
+    public int mapProductToOwners(Product product, Owner... owners) {
+        int count = 0;
+
+        for (Owner owner : owners) {
+            if (this.mapProductToOwner(product, owner)) {
+                ++count;
+            }
+        }
+
+        return count;
+    }
+
+    @Transactional
+    public int mapOwnerToProducts(Owner owner, Product... products) {
+        int count = 0;
+
+        for (Product product : products) {
+            if (this.mapProductToOwner(product, owner)) {
+                ++count;
+            }
+        }
+
+        return count;
+    }
+
+    @Transactional
+    public boolean removeOwnerFromProduct(Product product, Owner owner) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.product.uuid = :product_uuid AND op.owner.id = :owner_id";
+
+        int rows = this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .setParameter("product_uuid", product.getUuid())
+            .executeUpdate();
+
+        return rows > 0;
+    }
+
+    @Transactional
+    public int clearOwnersForProduct(Product product) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.product.uuid = :product_uuid";
+
+        return this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("product_uuid", product.getUuid())
+            .executeUpdate();
+    }
+
+    @Transactional
+    public int clearProductsForOwner(Owner owner) {
+        String jpql = "DELETE FROM OwnerProduct op " +
+            "WHERE op.owner.id = :owner_id";
+
+        return this.getEntityManager()
+            .createQuery(jpql)
+            .setParameter("owner_id", owner.getId())
+            .executeUpdate();
+    }
+
+    /**
+     * Updates the product references currently pointing to the original product to instead point to
+     * the updated product for the specified owners.
+     * <p/></p>
+     * <strong>Warning:</strong> Hibernate does not gracefully handle situations where the data
+     * backing an entity changes via direct SQL or other outside influence. While, logically, a
```

Comment:
This happened during unit testing and on PostgreSQL. If an entity contained another embedded or nested entity that was updated/removed by one of the reference modifiers ([update|remove][Product|Content]References), Hibernate would either throw an exception (remove) or revert back to the original state that had long been overwritten and flushed away (update). The only way I could get it to stop behaving this way is by evicting the entity entirely and requerying it.